File: src/debug.ts
import { appendFileSync, writeFileSync } from "node:fs";

const debugChannels = new Set<string>();
let debugAll = false;
let headless = false;
let logFile: string | null = null;

function initDebug(): void {
	debugChannels.clear();
	debugAll = false;
	logFile = process.env.HEDDLE_DEBUG_FILE ?? null;
	const val = process.env.HEDDLE_DEBUG;
	if (!val) return;
	if (val === "1" || val === "true") {
		debugAll = true;
		return;
	}
	for (const ch of val.split(",")) debugChannels.add(ch.trim());
}
initDebug();

/** Re-read HEDDLE_DEBUG and HEDDLE_DEBUG_FILE env vars and reset state. Useful for testing. */
export function resetDebug(): void {
	initDebug();
}

export function setHeadless(value: boolean): void {
	headless = value;
}

function formatArgs(args: unknown[]): string {
	return args
		.map((a) => (typeof a === "string" ? a : JSON.stringify(a)))
		.join(" ");
}

export function debug(channel: string, ...args: unknown[]): void {
	if (!debugAll && !debugChannels.has(channel)) return;
	const prefix = `[heddle:${channel}]`;
	if (logFile) {
		const timestamp = new Date().toISOString();
		appendFileSync(logFile, `${timestamp} ${prefix} ${formatArgs(args)}\n`);
	} else if (headless) {
		console.error(prefix, ...args);
	} else {
		console.debug(prefix, ...args);
	}
}

/** Clear the log file if one is configured. No-op otherwise. */
export function clearLogFile(): void {
	if (logFile) writeFileSync(logFile, "");
}

File: src/debug.rs
//! Channel-based debug logging — mirrors `ts-src/debug.ts`.
//!
//! Channels are configured via the `HEDDLE_DEBUG` env var (comma-separated list,
//! or `1`/`true` for everything). Output goes to a file if `HEDDLE_DEBUG_FILE`
//! is set, otherwise stderr in headless mode and stdout otherwise.

use std::collections::HashSet;
use std::fs::OpenOptions;
use std::io::Write;
use std::sync::RwLock;

use chrono::Utc;
use once_cell::sync::Lazy;

#[derive(Default)]
struct DebugState {
    channels: HashSet<String>,
    debug_all: bool,
    headless: bool,
    log_file: Option<String>,
}

static STATE: Lazy<RwLock<DebugState>> = Lazy::new(|| RwLock::new(load_state()));

fn load_state() -> DebugState {
    let mut state = DebugState::default();
    state.log_file = std::env::var("HEDDLE_DEBUG_FILE").ok();
    match std::env::var("HEDDLE_DEBUG") {
        Ok(val) if val == "1" || val == "true" => state.debug_all = true,
        Ok(val) if !val.is_empty() => {
            for ch in val.split(',') {
                state.channels.insert(ch.trim().to_string());
            }
        }
        _ => {}
    }
    state
}

/// Re-read `HEDDLE_DEBUG`/`HEDDLE_DEBUG_FILE` env vars. Used by tests.
pub fn reset_debug() {
    *STATE.write().unwrap() = load_state();
}

pub fn set_headless(value: bool) {
    STATE.write().unwrap().headless = value;
}

/// Emit a debug message on `channel`. No-op unless the channel (or `*`) is
/// enabled.
pub fn debug(channel: &str, msg: &str) {
    let state = STATE.read().unwrap();
    if !state.debug_all && !state.channels.contains(channel) {
        return;
    }
    let prefix = format!("[heddle:{channel}]");
    if let Some(log_path) = &state.log_file {
        let timestamp = Utc::now().to_rfc3339();
        if let Ok(mut file) = OpenOptions::new().append(true).create(true).open(log_path) {
            let _ = writeln!(file, "{timestamp} {prefix} {msg}");
        }
    } else if state.headless {
        eprintln!("{prefix} {msg}");
    } else {
        eprintln!("{prefix} {msg}");
    }
}

/// Truncate the log file if one is configured.
pub fn clear_log_file() {
    let state = STATE.read().unwrap();
    if let Some(path) = &state.log_file {
        let _ = std::fs::write(path, "");
    }
}

/// Build a debug message from a list of values. Convenience for places where
/// the TS code did `debug("ch", "msg:", obj)`.
pub fn debug_fmt(channel: &str, parts: &[&dyn std::fmt::Debug]) {
    let state = STATE.read().unwrap();
    if !state.debug_all && !state.channels.contains(channel) {
        return;
    }
    drop(state);
    let msg = parts
        .iter()
        .map(|p| format!("{p:?}"))
        .collect::<Vec<_>>()
        .join(" ");
    debug(channel, &msg);
}
